nupic/research/frameworks/pytorch/imagenet/mixins/lr_range_test.py

```diff
@@ -0,0 +1,111 @@
+# ------------------------------------------------------------------------------
+# Copyright (C) 2020, Numenta, Inc. All rights reserved.
+#
+# The information and source code contained herein is the
+# exclusive property of Numenta Inc.  No part of this software
+# may be used, reproduced, stored or distributed in any form,
+# without explicit written authorization from Numenta Inc.#
+# ------------------------------------------------------------------------------
+
+import os
+from copy import deepcopy
+
+from pandas import DataFrame
+from ray.tune.utils import flatten_dict
+
+from nupic.research.frameworks.pytorch.imagenet import mixins
+from nupic.research.frameworks.pytorch.lr_scheduler import LinearLRScheduler
+
+
+class LRRangeTest(mixins.LogEveryLoss):
+    """
+    Mixin for the LR-range test defined in section 4.1 of "A Disciplined Approach to
+    Neural Network Hyper-Parameters"
+        - https://arxiv.org/pdf/1803.09820.pdf
+
+    Herein, a min_lr and max_lr are set, and training proceeds for a small number of
+    epochs (1-3) while the learning rate is linearly increased. Generally, the point
+    at which the training loss begins to curve upwards and increases while the
+    validation/test loss still decreases, is considered to be a reasonable choice
+    for your max_lr in a cyclical lr-schedule. The same author recommends using 10-20
+    times lower this amount for your min_lr.
+    """
+
+    # Use the following name when `save_lr_test_to_csv=True`.
+    csv_save_name = "lr_range_test.csv"
+
+    def setup_experiment(self, config):
+        """
+        :param config:
+            - epochs: number of epochs in training (recommended 1-3)
+            - epochs_to_validate: will be overridden to include all epochs
+            - save_lr_test_to_csv: whether to save results to csv; must be accompanied
+                                   by 'logdir'
+            - logdir: directory to save test results
+            - lr_scheduler_class: automatically overridden to LinearLRScheduler
+            - lr_scheduler_args: args for the linear-schedule
+                - min_lr: starting learning rate
+                - max_lr: ending learning rate
+        """
+        # Ensure all epochs get validated.
+        assert "epochs" in config
+        config["epochs_to_validate"] = range(-1, config["epochs"])
+
+        # Init saving of test results.
+        self.save_lr_test_to_csv = config.get("save_lr_test_to_csv", False)
+        if self.save_lr_test_to_csv:
+            assert "logdir" in config, (
+                "The logdir must be specified to save the lt-range test resutls.")
+            self.logdir = config["logdir"]
+
+        # Save config for later - used to aggregate results.
+        self._config = deepcopy(config)
+        super().setup_experiment(config)
```

Comment:
Is there a reason you're calling super() at the end rather than the beginning? The convention is to always do it at the beginning unless you have a good reason not to. (e.g. profiling mixins)

Reply:
Fair point.